ADD: EIP-5489 NFT Hyperlink Extension (#5489)
ethereum/EIPs: EIPS/eip-5438.md

---
eip: 5438
title: NFT hyperlink extension
description: hyperlink extension for ERC-721
author: IronMan_CH (@IronMan_CH)
discussions-to: https://ethereum-magicians.org/t/eip-5438-NFT-hyperlink-extension/XXX
status: Draft
type: Standards Track
category: ERC
created: 2022-08-16

> requires: 165, 721

Author: 4907 is just taken as an example, I think no need to put it into requires.

---

## Abstract

> This ERC proposes a new extension of NFT (non-fungible token, aka ERC-721): nft-hyperlink-extention (hNFT), which embedded hyperlinks in NFTs. With hHNFT, the owners of NFTs can authorize a URL slot to a specific address which can be either an externally-owned account (EOA) or a contract account and they are entitled to revoke that authorization as well. Address which got slot authorization, can manage the URI on the slot.

Reviewer: ```suggestion
This ERC proposes a new extension of NFT (non-fungible token, aka [EIP-721](./eip-721.md)) that embeds hyperlinks in NFTs, referred to as "hNFT". With hHNFT, the owners of NFTs can authorize a URL slot to a specific address which can be either an externally-owned account (EOA) or a contract account and they are entitled to revoke that authorization as well. The address which has slot authorization can manage the URI on the slot.
```

## Motivation

> NFTs have entered mainstream consciousness in a big way as the trading volume spiked 21,000% to [more than $17 billion in 2021](https://www.cnbc.com/2022/03/10/trading-in-nfts-spiked-21000percent-to-top-17-billion-in-2021-report.html). Despite NFTs’ monumental rise as blockchain-based digital assets, we are just starting to scratch the surface of what is possible with NFTs.

Reviewer: ```suggestion
NFTs have entered mainstream consciousness in a big way as the trading volume spiked 21,000% to more than $17 billion in 2021. Despite NFTs’ monumental rise as blockchain-based digital assets, we are just starting to scratch the surface of what is possible with NFTs.
```

The emergence of NFTs brings more possibilities for creators and Web3 projects to build up decentralized communities, which also shifts the previous company-consumer relationship to more of a partnership. Each holder is entitled to trade, promote, or create based on their NFTs to make profits. 

However, there hasn’t been an effective approach to providing avenues of continued monetization for community holders. Going through all the NFT related EIPs, the solutions to ”making NFTs profitable“ are limited to transfers ([EIP-721](https://eips.ethereum.org/EIPS/eip-721)), rental and mortgage ([EIP 4907](https://eips.ethereum.org/EIPS/eip-4907)).

> Inspired by hyperlink, the essence of World Wide Web, hNFT standard is proposed positioning NFT as the medium for Web 3.0. HNFT standard endows NFTs with hyperlink extension so that it is possible to exercise the `fructus` right, meanwhile, keeps the `abusus`. With hyperlink extension, NFT owners can earn profit by authorizing any hyperlink slot to an address.

Author: Will do it.

> Inspired by hyperlink, the essence of World Wide Web, hNFT standard is proposed positioning NFT as the medium for Web 3.0. HNFT standard endows NFTs with hyperlink extension so that it is possible to exercise the `fructus` right, meanwhile, keeps the `abusus`. With hyperlink extension, NFT owners can earn profit by authorizing any hyperlink slot to an address.

Author: @Pandapip1  I have fixed the grammatical errors. Check it please.

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

### NFT hyperlink extension

NFT hyperlink extension extends ERC-721 with a `mapping<TokenId, mapping<SlotManagerAddress, uri>>`. The mapping means `TokenId` identified NFT has authorized `SlotManagerAddress` to manage the `uri` which is identified by `SlotManagerAddress`.

Besides, extension also provide the supporting facilities, such as
1. authorizeSlot/revokeSlot call pair used to manage slot authorization, and SlotAuthorizationCreate/Revoked event pair used to announce authorization management happens.
2. setSlotUri/getSlotUri call pair used to manage/view uri on slot, and SlotUriUpdated used to announce uri management happens.

### New URI format

This extension tries to add a new URI format, which is prefixed with "hnft://". This URI format enables redirect or connect activity between hNFTs, it enables great imagination.

### Interface

> #### IERC5438

Reviewer: ```suggestion
#### `IERC5438`
```

> ```

Reviewer: ```suggestion
```solidity
```

> //SPDX-License-Identifier: Unlicense

Reviewer: ```suggestion
// SPDX-License-Identifier: CC0-1.0
```

    ) external;

    /**
     * @dev
     * returns the latest uri of an slot on a token, which is indicated by `tokenId`, `slotManagerAddr`
     */
    function getSlotUri(uint256 tokenId, address slotManagerAddr)
        external
        view
        returns (string memory);
}
```

The `authorizeSlotTo(uint256 tokenId, address slotManagerAddr)` function MAY be implemented as public or external.

The `revokeAuthorization(uint256 tokenId, address slotManagerAddr)` function MAY be implemented as public or external.

The `revokeAllAuthorizations(uint256 tokenId)` function MAY be implemented as public or external.

The `setSlotUri(uint256 tokenId, string calldata newUri)` function MAY be implemented as public or external.

The `getSlotUri(uint256 tokenId, address slotManagerAddr)` function MAY be implemented as pure or view.

The `SlotAuthorizationCreated` event MUST be emitted when a slot is authorized to an address.

The `SlotAuthorizationRevoked` event MUST be emitted when a slot authorization is revoked.

The `SlotUriUpdated` event MUSt be emitted when a slot's URI is changed.

The `supportInterface` method MUST return true when called with `0x8f65987b`.

### Authentication

> The authorizeSlotTo, revokeAuthorization and revokeAllAuthorizations function is authenticated if and only if the message sender is the owner of the token.

Reviewer: ```suggestion
The `authorizeSlotTo`, `revokeAuthorization`, and `revokeAllAuthorizations` functions are authenticated if and only if the message sender is the owner of the token.
```

## Rationale

### Extends NFT with hyperlinks
We use URI to represent value of slot to ensure it's flexible enough to deal with different use cases. E.g. one can define a custom uri format to indicate a brand new protocol.

### Authorize slot to address
We use Address to represent key of slot to ensure it's flexible enough to deal with different use cases. E.g. one can manage slot's URI with an EOA, or manage slot's URI with a contract in which can do more detailed limitation or agreement.

### New URI format to enable redirect between hNFTs
This extension tries to add a new URI format, which is prefixed with `hnft://`. This URI format enables redirect or connect activity between hNFTs, it enables great imagination.

## Backwards Compatibility

> As mentioned in the specifications section, this standard can be fully ERC-721 compatible by adding an extension function set.

Reviewer: ```suggestion
As mentioned in the specifications section, this standard can be fully EIP-721 compatible by adding an extension function set.
```

> In addition, new functions introduced in this standard have many similarities with the existing functions in ERC-721. This allows developers to easily adopt the standard quickly.

Reviewer: ```suggestion
In addition, new functions introduced in this standard have many similarities with the existing functions in EIP-721. This allows developers to easily adopt the standard quickly.
```

## Reference Implementation

> You can find an implementation of this standard in [ERC5438.sol](../assets/eip-5438/contracts/ERC5438.sol).

Reviewer: ```suggestion
You can find an implementation of this standard in [`ERC5438.sol`](../assets/eip-5438/contracts/ERC5438.sol).
```

## Security Considerations
No security considerations were found.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

> 
> ## Citation
> Please cite this document as:
> 
> Ironman, "EIP-5438: hNFT, ERC-721 Hyperlink Extension," Ethereum Improvement Proposals, no. 5438, August 2022. [Online serial]. Available: https://eips.ethereum.org/EIPS/eip-5438.

Reviewer: ```suggestion
```
